tests/integration/test_integration.py

```diff
@@ -17,325 +18,331 @@
+    dash_duo.clear_input(input1)
+    input1.send_keys('hello world')
+
+    dash_duo.wait_for_text_to_equal('#output-1', 'hello world')
+    dash_duo.percy_snapshot(name='wildcard-callback-2')
+
+    # an initial call, one for clearing the input
+    # and one for each hello world character
+    assert input_call_count.value == 2 + len('hello world')
+
+    assert not dash_duo.get_logs()
+
+
+def test_inin003_aborted_callback(dash_duo):
+    """
+    Raising PreventUpdate OR returning no_update
+    prevents update and triggering dependencies
+    """
+
+    initial_input = 'initial input'
+    initial_output = 'initial output'
+
+    app = Dash(__name__)
+    app.layout = html.Div([
+        dcc.Input(id='input', value=initial_input),
+        html.Div(initial_output, id='output1'),
+        html.Div(initial_output, id='output2'),
+    ])
+
+    callback1_count = Value('i', 0)
+    callback2_count = Value('i', 0)
+
+    @app.callback(Output('output1', 'children'), [Input('input', 'value')])
+    def callback1(value):
+        callback1_count.value += 1
+        if callback1_count.value > 2:
+            return no_update
+        raise PreventUpdate("testing callback does not update")
+        return value
+
+    @app.callback(
+        Output('output2', 'children'), [Input('output1', 'children')]
+    )
+    def callback2(value):
+        callback2_count.value += 1
+        return value
+
+    dash_duo.start_server(app)
+
+    input_ = dash_duo.find_element('#input')
+    input_.send_keys('xyz')
+    dash_duo.wait_for_text_to_equal('#input', 'initial inputxyz')
+
+    # callback1 runs 4x (initial page load and 3x through send_keys)
+    until(lambda: callback1_count.value == 4, 3)
+
+    # callback2 is never triggered, even on initial load
+    assert callback2_count.value == 0
+
+    # double check that output1 and output2 children were not updated
+    assert dash_duo.find_element('#output1').text == initial_output
+    assert dash_duo.find_element('#output2').text == initial_output
+
+    assert not dash_duo.get_logs()
+
+    dash_duo.percy_snapshot(name='aborted')
+
+
+def test_inin004_wildcard_data_attributes(dash_duo):
+    app = Dash()
+    test_time = datetime.datetime(2012, 1, 10, 2, 3)
+    test_date = datetime.date(test_time.year, test_time.month,
+                              test_time.day)
+    app.layout = html.Div([
+        html.Div(
+            id="inner-element",
+            **{
+                'data-string': 'multiple words',
+                'data-number': 512,
+                'data-none': None,
+                'data-date': test_date,
+                'aria-progress': 5
+            }
+        )
+    ], id='data-element')
+
+    dash_duo.start_server(app)
+
+    div = dash_duo.find_element('#data-element')
+
+    # React wraps text and numbers with e.g. <!-- react-text: 20 -->
+    # Remove those
+    comment_regex = r'<!--[^\[](.*?)-->'  # noqa: W605
+
+    # Somehow the html attributes are unordered.
```

Comment:
this permutation was bad idea, I did similar clean up with beautiful soup parser for the initial state test case

Reply:
yeah I noticed that - `BeautifulSoup` is a nice idea, I'll try it here.